### ISSUE TITLE
HIP includes should have stronger include guards

### ISSUE THREAD

**Issue by robertmaynard** [kokkos/kokkos#3391]
Currently in `Kokkos_Macros.hpp` the inclusion of `<hip/hip_runtime.h>` and `<hip/hip_runtime_api.h>` occurs when `KOKKOS_ENABLE_HIP` was defined. As this define is set by the build-system it doesn't consider the fact that the consumer of the header might not be a hip compiler.

I encountered this when trying to compile a `c++` source that brought included a kokkos header but didn't use kokkos.

This will also become a more prevalent issue once CMake support HIP as a first class language

**Comment by crtrott** [comment]
Generally speaking as soon as you include any Kokkos header and you do enable a backend, it is that language period. Kokkos Macros for example does define KOKKOS_INLINE_FUNCTION as __device__ __host__ inline in HIP and CUDA. Hence as I said if you enble HIP ANY Kokkos header file is using the HIP language. 

Generally (since I have you here) we think its a bad idea to treat HIP as a language. At least if you want to distinguish languages by extension only or cmake doesn't allow you to "inject" a language via a header file. While it made quite a bit of sense in a world of separate compilation, for stuff like HIP or SYCL that feels just like the wrong approach. I'd rather have it treated like OpenMP.

In C++ world you will get more and more of these language dialects injected via header files. For example coming from the C++17 and beyond parallel algorithms. As a user of those I don't necessarily care what the underlying dialect is and thus wouldn't necessarily know which language to set for my files. If this is more like OpenMP it would actually work much easier.

As Kokkos we don't have a way of making a transitive language setting for our library our users inherit. RIght now. Furthermore, do I as a user have now to somehow make sure all the CMAKE_CXX_FLAGS get injected into CMAKE_CUDA_FLAGS, CMAKE_HIP_FLAGS, CMAKE_SYCL_FLAGS, CMAKE_OPENMPOFFLOAD_FLAGS, CMAKE_PLEASE_I_NEED_ANOTHER_VENDOR_MODEL_SINCE_ITS_SO_MUCH_FUN_AND_I_NEED_ANOTHER_30_YEARS_JOB_SECURITY_FLAGS?

This is essentially why we lie to CMAKE and tell it everything is C++ and we provide the necessary flags transitively on our own.

**Comment by crtrott** [comment]
Just to make that clear: adding a concept of "C++ dialekt" where you get both CMAKE_CXX_FLAGS and CMAKE_DIALEKT_FLAGS added to the compilation line, and where the dialekt is "infective", i.e. if I make it a public property of my library anything using it will also become that "C++ dialekt" would probably a pretty good solution for us, much better than either our current workaround or the current way CUDA is treated like Fortran as a different language from C++. And if my header files don't inject the dialect into compilation units which include them, then I could make it a private property.

**Comment by robertmaynard** [comment]
> Generally speaking as soon as you include any Kokkos header and you do enable a backend, it is that language period. Kokkos Macros for example does define KOKKOS_INLINE_FUNCTION as device host inline in HIP and CUDA. Hence as I said if you enble HIP ANY Kokkos header file is using the HIP language.

The specific issue I am trying to bring up is that the textual inclusion of `Kokkos_Macros.hpp` is broken for HIP in a subtly different way compared to CUDA.

For both CUDA and HIP I would expect that inclusion of usage of any macros defined in `Kokkos_Macros.hpp` by a C++ compiler would lead to errors. The problem is that for HIP  the `__host__` and `__device__` keywords/attributes aren't defined by the compiler but instead defined by the inclusion of `<hip/hip_runtime.h>`. 

So projects that currently use kokkos downstream via makefile / cmake / batch scripts and want to use different compilers for components now have a new issue where `kokkos` implicitly presumes that the HIP runtime includes can be found without any extra information by all compilers, which isn't the case.

**Comment by robertmaynard** [comment]
> In C++ world you will get more and more of these language dialects injected via header files. For example coming from the C++17 and beyond parallel algorithms. As a user of those I don't necessarily care what the underlying dialect is and thus wouldn't necessarily know which language to set for my files. If this is more like OpenMP it would actually work much easier.

To the best of my knowledge C++17 parallel based algorithms aren't going to be guarded behind a new language compilation flag by gcc or clang. In short, if you want cmake to treat CUDA and HIP as C++ you will need to have compilers such as clang to drop `-x hip` and `-x cuda` and infer those language constructs some other way.

**Comment by crtrott** [comment]
But how do you include Kokkos_Macros.hpp without getting all the Kokkos provided flags? I mean are you trying to include individual Kokkos flags wihtout depending on the Kokkos target? That sounds very iffy.

**Comment by robertmaynard** [comment]
Kokkos doesn't provide the include directory to `hip/hip_runtime.h` via any of the targets or flags. It is either injected implicitly by the `hipcc` wrapper, or required to be provided to clang when using `clang -x hip`.

Since the C++ compiler isn't either of the above it has no knowledge of what include would be sufficient to find that header.

**Comment by crtrott** [comment]
yeah but the C++ compiler would anyway fail to compile the file, are you saying the problem is we need to provide it if clang -x hip is used as compiler?

**Comment by crtrott** [comment]
I.e. if the issue is that we only support hipcc as HIP compiler, than the answer is we need to maybe also support native clang and provide the necessary flags on our own.

**Comment by robertmaynard** [comment]
> yeah but the C++ compiler would anyway fail to compile the file, are you saying the problem is we need to provide it if clang -x hip is used as compiler?

I am saying that simply doing `#include` on a Kokkos header isn't the same as usage of the defines in that kokkos header. In this case we are textually including a Kokkos header with no intent to use anything it defines, and I was expecting that to work. 

The correct solution might be for kokkos not to use `__host__` / `__device__` and instead use `__attribute__((host))` and `__attribute__((device))` ( ignoring hcc ) as those are what the macros evaluate to and don't require the inclusion of any hip headers.

**Comment by pfultz2** [comment]
The hip_runtime.h header is meant for GPU compilation so it requires a device compiler(ie clang or hcc), whereas hip_runtime_api.h can be used from the host and with any compiler. If the public headers are including hip_runtime.h then it should link with the cmake target `hip::device` in its interface, and all downstream users will be required to use clang.

**Comment by robertmaynard** [comment]
What about kokkos makefile users?

**Comment by masterleinad** [comment]
https://github.com/kokkos/kokkos/pull/3327#issuecomment-685836194:
> Also we do not support installing Kokkos via Makefile.kokkos, which means the only thing you can do with the Makefile is building an app directly with Kokkos as an inline build [...].

**Comment by pfultz2** [comment]
> What about kokkos makefile users?

For now, you will need to use hipcc, as hip doesn't have support for pkg-config yet.

**Comment by robertmaynard** [comment]
The proposed solution that `kokkos::kokkoscore` or `kokkos::kokkos` will export a dependency on `hip::host` seems reasonable to me. 

I am happy to have this issue closed, as in the short-term I can make sure our downstream find_package(Kokkos) adjusts the interface to rely on `hip:host`

**Comment by PhilMiller** [comment]
@crtrott I think we can close this now?
